assets/src/edit-story/components/checklist/checkpointContext/checkpointContext.js

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * External dependencies
+ */
+import { useEffect, useState, useCallback, useReducer, useMemo } from 'react';
+import PropTypes from 'prop-types';
+import {
+  createContext,
+  useContextSelector,
+  identity,
+} from '@web-stories-wp/design-system';
+/**
+ * Internal dependencies
+ */
+import { STORY_EVENTS, useStory, useStoryTriggerListener } from '../../../app';
+import { PPC_CHECKPOINT_STATE } from '../constants';
+
+export const PPC_CHECKPOINT_ACTION = {
+  ON_INITIAL_ELEMENT_ADDED: 'story is no longer empty',
+  ON_PUBLISH_CLICKED: 'publish button is pressed',
+  ON_STORY_HAS_2_PAGES: "story 'recommended' section enabled",
+  ON_STORY_HAS_5_PAGES: "story 'high priority' section enabled",
+  ON_STORY_IS_PUBLISHED: 'story is published, regardless of recommendations',
+};
+
+const machine = {
+  [PPC_CHECKPOINT_STATE.UNAVAILABLE]: {
+    [PPC_CHECKPOINT_ACTION.ON_INITIAL_ELEMENT_ADDED]:
+      PPC_CHECKPOINT_STATE.ONLY_RECOMMENDED,
+    [PPC_CHECKPOINT_ACTION.ON_PUBLISH_CLICKED]: PPC_CHECKPOINT_STATE.ALL,
+    [PPC_CHECKPOINT_ACTION.ON_STORY_IS_PUBLISHED]: PPC_CHECKPOINT_STATE.ALL,
+    [PPC_CHECKPOINT_ACTION.ON_STORY_HAS_2_PAGES]:
+      PPC_CHECKPOINT_STATE.ONLY_RECOMMENDED,
+  },
+  [PPC_CHECKPOINT_STATE.ONLY_RECOMMENDED]: {
+    [PPC_CHECKPOINT_ACTION.ON_STORY_HAS_5_PAGES]: PPC_CHECKPOINT_STATE.ALL,
+    [PPC_CHECKPOINT_ACTION.ON_PUBLISH_CLICKED]: PPC_CHECKPOINT_STATE.ALL,
+    [PPC_CHECKPOINT_ACTION.ON_STORY_IS_PUBLISHED]: PPC_CHECKPOINT_STATE.ALL,
+  },
+  [PPC_CHECKPOINT_STATE.ALL]: {},
+};
+
+const checkpointReducer = (state, action) => {
+  return machine[state][action] || state;
+};
+
+export const CheckpointContext = createContext({ state: {}, actions: {} });
+
+function ChecklistCheckpointProvider({ children }) {
+  const [checkpointState, dispatch] = useReducer(
+    checkpointReducer,
+    PPC_CHECKPOINT_STATE.UNAVAILABLE
+  );
+
+  const [highPriorityCount, setHighPriorityCount] = useState();
```

Comment:
🤔 This feels like redundant state because this count is already tracked in the [`ChecklistCountProvider`](https://github.com/google/web-stories-wp/blob/main/assets/src/edit-story/components/checklist/checkCountContext.js)

The `useCategoryCount` hook would be good to use here, but I think that provider is a child of this provider. I think it'd be ideal to combine that state though

Reply:
This is a tricky bit because I need the high priority count to tell me if that review dialog should pop up on 'publish' but that count isn't something i can use the hook for in this provider because, like you said, it's a child of this one. So I was making due by using `setHighPriority` in that `updateHighPriorityCount` where both hooks are available to hoist the count to the checkpoint and decided if the dialog should be seen.

Reply:
Yeah that makes sense. It seems like a huge lift (especially with the already big pr) to move the count provider up higher so that this context and the rest of the PPC have access to the same state.

For now let's leave it and then if we ever need to do this again with the design/accessibility counts or any other state then let's lift it 😄

Reply:
deal! 🤝